File: java/Day11.java
import java.io.*;
import java.math.BigInteger;
import java.util.*;

public class Day11 {
    static Map<Integer, Monkey> map;
    static int MOD = 1;
    
    public static void main(String[] args) throws FileNotFoundException {
        Scanner in = new Scanner(new FileReader("input/day11.txt"));
        map = new TreeMap<>();
        int number = 0;
        while (in.hasNext()) {
            in.nextLine();
            String[] itemsStr = in.nextLine().split(": ")[1].split(", ");
            Queue<BigInteger> items = new LinkedList<>();
            for (String s : itemsStr) {
                items.offer(new BigInteger(s));
            }
            
            // [operator, variable]
            String[] operation = in.nextLine().split("= old ")[1].split(" ");
            int test = getLastDigit(in.nextLine());
            MOD *= test;
            int[] next = {getLastDigit(in.nextLine()), getLastDigit(in.nextLine())};
            
            Monkey monkey = new Monkey(number, items, operation, test, next, 0);
            map.put(number, monkey);
            if (in.hasNext()) in.nextLine();
            number++;
        }
        for (int i = 0; i < 10000; i++) {
            inspect();
        }
        System.out.println("Most active: " + twoMostActive());
        // 99852, 25935263541
    }
    
    // ranking activity
    private static long twoMostActive() {
        Queue<Long> pq = new PriorityQueue<>();
        for (Monkey monkey : map.values()) {
            pq.add((long) monkey.inspected);
            if (pq.size() > 2) pq.poll();
        }
        return pq.poll() * pq.poll();
    }
    
    // apply inspections
    private static void inspect() {
        for (Monkey monkey : map.values()) {
            while (!monkey.items.isEmpty()) {
                BigInteger item = operate(monkey.items.poll(), monkey.operation).mod(BigInteger.valueOf(MOD));
                BigInteger test = BigInteger.valueOf(monkey.test);
                int nextMonkey = monkey.next[item.mod(test).equals(BigInteger.valueOf(0)) ? 0 : 1];
                map.get(nextMonkey).items.offer(item);
                monkey.inspected++;
            }
        }
    }
    
    // apply operations
    private static BigInteger operate(BigInteger item, String[] operation) {
        BigInteger var = operation[1].equals("old") ? item : new BigInteger(operation[1]);
        if (operation[0].equals("*")) {
            return item.multiply(var);
        } else {
            return item.add(var);
        }
    }
    
    private static int getLastDigit(String line) {
        String[] str = line.split(" ");
        return Integer.parseInt(str[str.length - 1]);
    }
    
    // Monkey object with all the info
    static class Monkey {
        int number;
        Queue<BigInteger> items;
        String[] operation;
        int test;
        int[] next;
        int inspected;
        
        private Monkey(int number, Queue<BigInteger> items, String[] operation, int test, int[] next, int inspected) {
            this.number = number;
            this.items = items;
            this.operation = operation;
            this.test = test;
            this.next = next;
            this.inspected = inspected;
        }
    }
}

File: py/Day11.py
import heapq

class Monkey:
    def __init__(self, number, items, operation, multiplier, test, next, inspected):
        self.number = number
        self.items = items
        self.operation = operation
        self.multiplier = multiplier
        self.test = test
        self.next = next
        self.inspected = inspected

map = {}
mod = 1

def main():
    global mod
    num = 0
    with open('../input/day11.txt') as file:
        while file.readline():
            items = [int(n) for n in file.readline().strip().split(": ")[1].split(", ")]
            heapq.heapify(items)
            operation, multiplier = file.readline().strip().split("= old ")[1].split(" ")
            test = int(file.readline().strip().split(" ")[-1])
            mod *= test
            next = [int(file.readline().strip().split(" ")[-1]) for _ in range(2)]
            map[num], num = Monkey(num, items, operation, multiplier, test, next, 0), num + 1
            file.readline()

    for _ in range(10000):
        inspect()
    print("25935263541 =", two_most_active())

def two_most_active():
    pq = []
    heapq.heapify(pq)
    [heapq.heappop(pq) if len(pq) > 2 else heapq.heappush(pq, monkey.inspected) for monkey in map.values()]
    return heapq.heappop(pq) * heapq.heappop(pq)

def inspect():
    for monkey in map.values():
        while monkey.items:
            item = operate(heapq.heappop(monkey.items), monkey.operation, monkey.multiplier) % mod
            heapq.heappush(map[monkey.next[item % monkey.test != 0]].items, item)
            monkey.inspected += 1

def operate(item, operation, multiplier):
    var = item if multiplier == "old" else int(multiplier)
    return item * var if operation == "*" else item + var

if __name__ == "__main__":
    main()
